### PR TITLE
ICU-20777 Add test to validate XLikelySubtags

### PR DESCRIPTION
**Pull request by FrankYFTang**
We need to first validate that XLikelySubtags implemented the algorithm correctly before switching over to it.

##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-20777
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [X] Issue accepted (done by Technical Committee after discussion)
- [X] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Review by macchiati — COMMENTED** [review]
Looks good to me (but my C/C++ is very rusty). One comment is that the test cases should be moved into a CLDR testData file someday.

File: icu4c/source/common/lsr.h [hunk 1]
```diff
@@ -45,7 +45,12 @@ struct LSR final : public UMemory {
      */
     LSR(char prefix, const char *lang, const char *scr, const char *r, int32_t f,
         UErrorCode &errorCode);
-    LSR(LSR &&other) noexcept;
+    /**
+     * Constructor which copies those into owned memory, and aliases the region.
+     */
+    LSR(const char *lang, const char *scr, const char *r, int32_t f,
+        UErrorCode &errorCode);
+    LSR(LSR &&othe) noexcept;
```
Comment (pedberg-icu) [inline icu4c/source/common/lsr.h:53]: Shouldn't this have &&other ?

**Comment by pedberg-icu** [comment]
Hi Frank, can you explain (and maybe add in a comment in th test) how the goal of this test is different from e.g. LocaleTest::TestAddLikelyAndMinimizeSubtags, LocaleTest::TestAddLikelySubtags, LocaleTest::TestMinimizeSubtags in lockets.cpp? Thanks.

**Comment by FrankYFTang** [comment]
> TestAddLikelyAndMinimizeSubtags

**Comment by FrankYFTang** [comment]
> Hi Frank, can you explain (and maybe add in a comment in th test) how the goal of this test is different from e.g. LocaleTest::TestAddLikelyAndMinimizeSubtags, LocaleTest::TestAddLikelySubtags, LocaleTest::TestMinimizeSubtags in lockets.cpp? Thanks.

LocaleTest is testing against the methods in Locale class.

**Comment by FrankYFTang** [comment]
> Looks good to me (but my C/C++ is very rusty). One comment is that the test cases should be moved into a CLDR testData file someday.

ok, how about I rewrite the tests to make add these cases to  LocaleTest::TestAddLikelySubtags and also change the code inside  LocaleTest::TestAddLikelySubtags to ALSO call XLikelySubtags ?

**Comment by macchiati** [comment]
Sgtm

**Comment by FrankYFTang** [comment]
right now, we got the following erros:
```
   TestAddLikelyAndMinimizeSubtags {
         FAIL: makeMaximizedLsrFrom(und_Ethi_ER).language; got "ti"; expected "am"
         FAIL: makeMaximizedLsrFrom(und_Latn_HK).language; got "en"; expected "zh"
         FAIL: makeMaximizedLsrFrom(art).script; got "Latn"; expected ""
         FAIL: makeMaximizedLsrFrom(art).region; got "US"; expected ""
         FAIL: makeMaximizedLsrFrom(art_ZZ).script; got "Latn"; expected ""
         FAIL: makeMaximizedLsrFrom(art_ZZ).region; got "US"; expected "ZZ"
         FAIL: makeMaximizedLsrFrom(art_CN).script; got "Hans"; expected ""
         FAIL: makeMaximizedLsrFrom(art_TW).script; got "Hant"; expected ""
         FAIL: makeMaximizedLsrFrom(art_HK).script; got "Hant"; expected ""
         FAIL: makeMaximizedLsrFrom(art_AQ).script; got "Latn"; expected ""
         FAIL: makeMaximizedLsrFrom(art_Zzzz).script; got "Latn"; expected "Zzzz"
         FAIL: makeMaximizedLsrFrom(art_Zzzz).region; got "US"; expected ""
         FAIL: makeMaximizedLsrFrom(art_Zzzz_ZZ).script; got "Latn"; expected "Zzzz"
         FAIL: makeMaximizedLsrFrom(art_Zzzz_ZZ).region; got "US"; expected "ZZ"
         FAIL: makeMaximizedLsrFrom(art_Zzzz_CN).script; got "Hans"; expected "Zzzz"
         FAIL: makeMaximizedLsrFrom(art_Zzzz_TW).script; got "Hant"; expected "Zzzz"
         FAIL: makeMaximizedLsrFrom(art_Zzzz_HK).script; got "Hant"; expected "Zzzz"
         FAIL: makeMaximizedLsrFrom(art_Zzzz_AQ).script; got "Latn"; expected "Zzzz"
         FAIL: makeMaximizedLsrFrom(art_Latn).region; got "US"; expected ""
         FAIL: makeMaximizedLsrFrom(art_Latn_ZZ).region; got "US"; expected "ZZ"
         FAIL: makeMaximizedLsrFrom(art_Hans).region; got "CN"; expected ""
         FAIL: makeMaximizedLsrFrom(art_Hans_ZZ).region; got "CN"; expected "ZZ"
         FAIL: makeMaximizedLsrFrom(art_Hant).region; got "TW"; expected ""
         FAIL: makeMaximizedLsrFrom(art_Hant_ZZ).region; got "TW"; expected "ZZ"
         FAIL: makeMaximizedLsrFrom(art_Moon).region; got "US"; expected ""
         FAIL: makeMaximizedLsrFrom(art_Moon_ZZ).region; got "US"; expected "ZZ"
         FAIL: makeMaximizedLsrFrom(und_150).region; got "150"; expected "RU"
```

**Comment by macchiati** [comment]
Those are failures. Eg for the first, CLDR has 		<likelySubtag from="und_Ethi" to="am_Ethi_ET"/>
so und_Ethi => am_Ethi_ET
and und_Ethi_XX => am_Ethi_XX

**Comment by FrankYFTang** [comment]
but CLDR also have data
```
		<likelySubtag from="und_ER" to="ti_Ethi_ER"/>
```

UTS35 https://www.unicode.org/reports/tr35/#Likely_Subtags said
```
Lookup. Look up each of the following in order, and stop on the first match:
languages_scripts_regions
languages_regions
languages_scripts
languages
und_scripts
```

so if the input is und_Ethi_ER
1. it first search
und_Ethi_ER and found no match,
2. then search for
und_ER and found the match ti_Ethi_ER
and therefore maximize to ti_Ethi_ER
It has not yet try to search for und_Thai yet.

**Comment by macchiati** [comment]
Excellent point, you are right.

**Pull request closed without merging**